runtime/src/accounts_index.rs

```diff
@@ -2064,22 +1984,66 @@ pub mod tests {
             // clone the AccountMapEntryInner into a new Arc
             match self {
                 PreAllocatedAccountMapEntry::Entry(entry) => {
-                    let (slot, account_info) = entry.slot_list.read().unwrap()[0];
+                    let (slot, account_info) = entry.slot_list[0];
                     let meta = AccountMapEntryMeta {
                         dirty: AtomicBool::new(entry.dirty()),
                         age: AtomicU8::new(entry.age()),
                     };
-                    PreAllocatedAccountMapEntry::Entry(Arc::new(AccountMapEntryInner::new(
+                    PreAllocatedAccountMapEntry::Entry(AccountMapEntryInner::new(
                         vec![(slot, account_info)],
                         entry.ref_count(),
                         meta,
-                    )))
+                    ))
                 }
                 PreAllocatedAccountMapEntry::Raw(raw) => PreAllocatedAccountMapEntry::Raw(*raw),
             }
         }
     }
 
+    impl<T: IndexValue> Clone for AccountMapEntry<T> {
+        fn clone(&self) -> Self {
+            Self {
+                slot_list: self.slot_list().clone(),
+                ref_count: self.ref_count,
+                meta: AccountMapEntryMeta {
+                    dirty: AtomicBool::new(self.dirty()),
+                    age: AtomicU8::new(self.age()),
+                },
+            }
+        }
+    }
+
+    impl<T: Clone> AccountMapEntry<T> {
+        pub fn new(slot_list: SlotList<T>, ref_count: RefCount, meta: AccountMapEntryMeta) -> Self {
+            Self {
+                slot_list,
+                ref_count,
+                meta,
+            }
+        }
+    }
+
+    impl<T: IndexValue> AccountsIndex<T> {
+        pub fn get(
+            &self,
+            pubkey: &Pubkey,
+            ancestors: Option<&Ancestors>,
+            max_root: Option<Slot>,
+        ) -> Option<(AccountMapEntry<T>, usize)> {
+            self.get2(pubkey, ancestors, max_root, |entry| {
+                entry.map(|(entry, index)| (entry.clone(), index))
```

Comment:
Noting for deeper review later.

Previous implementation of `get` could return a `AccountIndexGetResult::Found` which would hold the read lock to the caller, whereas this change only clones the AccountMapEntry, meaning the backing index entry could have been modified when this function returns to the caller.

TODO: Check every caller/entrypoint to `AccountsIndex::get()` to make sure this new change doesn't break assumptions